File: scripts/keep_alive.py
"""
Zeabur Keep Alive Script (Python)
使用 Playwright 模拟浏览器登录，保持账户活跃
登录成功后发送 Telegram 通知和截图
"""

import os
import sys
from datetime import datetime
from playwright.sync_api import sync_playwright
from update_secret import update_secret
from telegram_notify import send_telegram_message, send_telegram_photo

ZEABUR_DASHBOARD_URL = 'https://zeabur.com/projects'
SCREENSHOT_PATH = '/tmp/zeabur_dashboard.png'


def main():
    cookie_string = os.environ.get('ZEABUR_COOKIE')
    repo_token = os.environ.get('REPO_TOKEN')
    repo = os.environ.get('GITHUB_REPOSITORY', '')
    tg_bot_token = os.environ.get('TG_BOT_TOKEN')
    tg_chat_id = os.environ.get('TG_CHAT_ID')

    if not cookie_string:
        print('❌ 错误: ZEABUR_COOKIE 环境变量未设置')
        sys.exit(1)

    print('🚀 启动浏览器...')
    
    with sync_playwright() as p:
        browser = p.chromium.launch(headless=True)
        context = browser.new_context()
        
        # 解析并设置 Cookie
        cookies = parse_cookies(cookie_string)
        context.add_cookies(cookies)
        
        page = context.new_page()
        
        try:
            print('📡 访问 Zeabur 控制台...')
            page.goto(ZEABUR_DASHBOARD_URL, wait_until='networkidle')
            
            # 检查登录状态
            is_logged_in = check_login_status(page)
            
            if not is_logged_in:
                error_msg = '❌ 登录失败: Cookie 可能已过期'
                print(error_msg)
                if tg_bot_token and tg_chat_id:
                    send_telegram_message(tg_bot_token, tg_chat_id, error_msg)
                sys.exit(1)
            
            now = datetime.now().strftime('%Y-%m-%d %H:%M:%S')
            print(f'✅ 登录成功！')
            print(f'⏰ 执行时间: {now}')
            
            # 截图
            page.screenshot(path=SCREENSHOT_PATH, full_page=False)
            print(f'📸 截图已保存: {SCREENSHOT_PATH}')
            
            # 发送 Telegram 通知
            if tg_bot_token and tg_chat_id:
                message = f'✅ Zeabur 保活成功！\n⏰ 时间: {now}'
                send_telegram_message(tg_bot_token, tg_chat_id, message)
                send_telegram_photo(tg_bot_token, tg_chat_id, SCREENSHOT_PATH, caption='Zeabur 控制台截图')
                print('📤 Telegram 通知已发送')
            
            # 提取并更新 Cookie
            new_cookies = context.cookies()
            new_cookie_string = format_cookies(new_cookies)
            
            if repo_token and repo and new_cookie_string != cookie_string:
                print('🔄 检测到 Cookie 变化，正在更新 GitHub Secret...')
                owner, repo_name = repo.split('/')
                update_secret(repo_token, owner, repo_name, 'ZEABUR_COOKIE', new_cookie_string)
                print('✅ GitHub Secret 已更新')
        
        except Exception as e:
            error_msg = f'❌ 执行失败: {str(e)}'
            print(error_msg)
            if tg_bot_token and tg_chat_id:
                send_telegram_message(tg_bot_token, tg_chat_id, error_msg)
            sys.exit(1)
        
        finally:
            browser.close()


def check_login_status(page) -> bool:
    """检查登录状态"""
    try:
        url = page.url
        if '/login' in url:
            return False
        
        page.wait_for_timeout(2000)
        title = page.title()
        return 'Zeabur' in title and 'Login' not in title
    except:
        return False


def parse_cookies(cookie_string: str) -> list:
    """解析 Cookie 字符串为 Playwright 格式"""
    cookies = []
    for cookie in cookie_string.split(';'):
        parts = cookie.strip().split('=', 1)
        if len(parts) == 2:
            name, value = parts
            cookies.append({
                'name': name.strip(),
                'value': value.strip(),
                'domain': '.zeabur.com',
                'path': '/',
            })
    return cookies


def format_cookies(cookies: list) -> str:
    """格式化 Cookies 为字符串"""
    return '; '.join(
        f"{c['name']}={c['value']}"
        for c in cookies
        if 'zeabur.com' in c.get('domain', '')
    )


if __name__ == '__main__':
    main()

File: scripts/keep-alive.js
/**
 * Zeabur Keep Alive Script
 * 使用 Playwright 模拟浏览器登录，保持账户活跃
 */

import { chromium } from 'playwright';
import { updateSecret } from './update-secret.js';

const ZEABUR_DASHBOARD_URL = 'https://zeabur.com/projects';

async function main() {
  const cookieString = process.env.ZEABUR_COOKIE;
  const repoToken = process.env.REPO_TOKEN;
  const repoOwner = process.env.GITHUB_REPOSITORY_OWNER;
  const repoName = process.env.GITHUB_REPOSITORY?.split('/')[1];

  if (!cookieString) {
    console.error('❌ 错误: ZEABUR_COOKIE 环境变量未设置');
    process.exit(1);
  }

  console.log('🚀 启动浏览器...');
  const browser = await chromium.launch({ headless: true });
  const context = await browser.newContext();

  // 解析并设置 Cookie
  const cookies = parseCookies(cookieString);
  await context.addCookies(cookies);

  const page = await context.newPage();

  try {
    console.log('📡 访问 Zeabur 控制台...');
    await page.goto(ZEABUR_DASHBOARD_URL, { waitUntil: 'networkidle' });

    // 检查是否登录成功（页面上应该有项目列表或用户信息）
    const isLoggedIn = await checkLoginStatus(page);

    if (!isLoggedIn) {
      console.error('❌ 登录失败: Cookie 可能已过期');
      process.exit(1);
    }

    console.log('✅ 登录成功！');
    console.log(`⏰ 执行时间: ${new Date().toISOString()}`);

    // 提取更新后的 Cookie
    const newCookies = await context.cookies();
    const newCookieString = formatCookies(newCookies);

    // 如果 Cookie 有变化且配置了更新参数，则更新 GitHub Secret
    if (repoToken && repoOwner && repoName && newCookieString !== cookieString) {
      console.log('🔄 检测到 Cookie 变化，正在更新 GitHub Secret...');
      await updateSecret(repoToken, repoOwner, repoName, 'ZEABUR_COOKIE', newCookieString);
      console.log('✅ GitHub Secret 已更新');
    }

  } catch (error) {
    console.error('❌ 执行失败:', error.message);
    process.exit(1);
  } finally {
    await browser.close();
  }
}

/**
 * 检查登录状态
 */
async function checkLoginStatus(page) {
  try {
    // 检查是否被重定向到登录页
    const url = page.url();
    if (url.includes('/login')) {
      return false;
    }

    // 等待页面加载完成，检查是否有项目相关内容
    await page.waitForTimeout(2000);
    
    // 检查页面标题或特定元素
    const title = await page.title();
    return title.includes('Zeabur') && !title.includes('Login');
  } catch {
    return false;
  }
}

/**
 * 解析 Cookie 字符串为 Playwright Cookie 格式
 * 支持格式: "name1=value1; name2=value2"
 */
function parseCookies(cookieString) {
  return cookieString.split(';').map(cookie => {
    const [name, ...valueParts] = cookie.trim().split('=');
    return {
      name: name.trim(),
      value: valueParts.join('=').trim(),
      domain: '.zeabur.com',
      path: '/',
    };
  }).filter(c => c.name && c.value);
}

/**
 * 格式化 Cookies 为字符串
 */
function formatCookies(cookies) {
  return cookies
    .filter(c => c.domain.includes('zeabur.com'))
    .map(c => `${c.name}=${c.value}`)
    .join('; ');
}

main();
